coderd/notifications/dispatch/smtp.go

```diff
@@ -0,0 +1,338 @@
+package dispatch
+
+import (
+	"bytes"
+	"context"
+	_ "embed"
+	"fmt"
+	"mime/multipart"
+	"mime/quotedprintable"
+	"net"
+	"net/mail"
+	"net/smtp"
+	"net/textproto"
+	"os"
+	"strings"
+	"time"
+
+	"github.com/google/uuid"
+	"golang.org/x/xerrors"
+
+	"cdr.dev/slog"
+	"github.com/coder/coder/v2/coderd/database"
+	"github.com/coder/coder/v2/coderd/notifications/render"
+	"github.com/coder/coder/v2/coderd/notifications/types"
+	"github.com/coder/coder/v2/codersdk"
+)
+
+var (
+	ValidationNoFromAddressErr   = xerrors.New("no 'from' address defined")
+	ValidationNoSmarthostHostErr = xerrors.New("smarthost 'host' is not defined, or is invalid")
+	ValidationNoSmarthostPortErr = xerrors.New("smarthost 'port' is not defined, or is invalid")
+	ValidationNoHelloErr         = xerrors.New("'hello' not defined")
+
+	//go:embed smtp/html.gotmpl
+	htmlTemplate string
+	//go:embed smtp/plaintext.gotmpl
+	plainTemplate string
+)
+
+// SMTPHandler is responsible for dispatching notification messages via SMTP.
+// NOTE: auth and TLS is currently *not* enabled in this initial thin slice.
+// TODO: implement auth
+// TODO: implement TLS
+type SMTPHandler struct {
+	cfg codersdk.NotificationsEmailConfig
+	log slog.Logger
+}
+
+func NewSMTPHandler(cfg codersdk.NotificationsEmailConfig, log slog.Logger) *SMTPHandler {
+	return &SMTPHandler{cfg: cfg, log: log}
+}
+
+func (*SMTPHandler) NotificationMethod() database.NotificationMethod {
+	return database.NotificationMethodSmtp
+}
+
+func (s *SMTPHandler) Dispatcher(payload types.MessagePayload, titleTmpl, bodyTmpl string) (DeliveryFunc, error) {
+	// First render the subject & body into their own discrete strings.
+	subject, err := render.Plaintext(titleTmpl)
+	if err != nil {
+		return nil, xerrors.Errorf("render subject: %w", err)
+	}
+
+	htmlBody, err := render.HTML(bodyTmpl)
+	if err != nil {
+		return nil, xerrors.Errorf("render HTML body: %w", err)
+	}
+
+	plainBody, err := render.Plaintext(bodyTmpl)
+	if err != nil {
+		return nil, xerrors.Errorf("render plaintext body: %w", err)
+	}
+
+	// Then, reuse these strings in the HTML & plain body templates.
+	payload.Labels.Set("_subject", subject)
+	payload.Labels.Set("_body", htmlBody)
+	htmlBody, err = render.GoTemplate(htmlTemplate, payload, nil)
+	if err != nil {
+		return nil, xerrors.Errorf("render full html template: %w", err)
+	}
+	payload.Labels.Set("_body", plainBody)
+	plainBody, err = render.GoTemplate(plainTemplate, payload, nil)
+	if err != nil {
+		return nil, xerrors.Errorf("render full plaintext template: %w", err)
+	}
+
+	return s.dispatch(subject, htmlBody, plainBody, payload.UserEmail), nil
+}
+
+// dispatch returns a DeliveryFunc capable of delivering a notification via SMTP.
+//
+// NOTE: this is heavily inspired by Alertmanager's email notifier:
+// https://github.com/prometheus/alertmanager/blob/342f6a599ce16c138663f18ed0b880e777c3017d/notify/email/email.go
+func (s *SMTPHandler) dispatch(subject, htmlBody, plainBody, to string) DeliveryFunc {
```

Comment:
nit: this could probably be extracted to its own package outside of `coderd`

Reply:
What use-case do you see for SMTP delivery outside of notifications?

Reply:
Password reset emails will be a notification as well.
When I add more tests for SMTP I'll see if extraction feels like an ergonomic step 👍